src/services/executionEnvironment.ts

```diff
@@ -0,0 +1,316 @@
+import * as child_process from 'child_process';
+import * as fs from 'fs';
+import * as path from 'path';
+import { URI } from 'vscode-uri';
+import { Connection } from 'vscode-languageserver';
+import { v4 as uuidv4 } from 'uuid';
+import { asyncExec } from '../utils/misc';
+import { WorkspaceFolderContext } from './workspaceManager';
+import { IContianerEngine } from '../interfaces/extensionSettings';
+
+
+export class ExecutionEnvironment {
+    private connection: Connection;
+    private context: WorkspaceFolderContext;
+    private _container_engine: IContianerEngine;
+    private _container_image: string;
+    private _container_image_id: string;
+
+    constructor(connection: Connection, context: WorkspaceFolderContext) {
+      this.connection = connection;
+      this.context = context;
+    }
+
+    public async initialize(): Promise<void> {
+      try {
+        const settings = await this.context.documentSettings.get(
+          this.context.workspaceFolder.uri
+        );
+        this._container_image = settings.executionEnvironment.image
+        this._container_engine = settings.executionEnvironment.containerEngine
+        if (this._container_engine === 'auto') {
+          for (const ce of ['podman', 'docker']) {
+              try {
```

Comment:
Try/except should only wrap one instruction to avoid surprises.

Reply:
In this case, after the try block is executing only function and the next line is an assignment statement.

Reply:
I don't think it'd increase the complexity, vice versa — it will decrease nesting for some of the code which will make it more linear when reading.

Reply:
I don't see how it will decrease nesting, in-fact I think the current structure is moe linear compared to introducing both continue and break.

Reply:
In this particalue we are pretty sure that there won't be any surprises as such

Reply:
Those instructions after the invocation could be dedented, and that's what could be more linear.

Besides, I'm pretty sure that `this.connection.console.log()` may error out on connection errors essentially breaking the flow. This will be hard to catch.

Reply:
Can you provide steps to reproduce the error thrown in `this.connection.console.log()`  for connection errors?